### ISSUE TITLE
avoid changes in the capitalization of file names

### ISSUE THREAD

**Issue by TP75** [splitbrain/dokuwiki#2359]
Please avoid name changes to required files in the capitalization of their letters only. This can lead to issues on case-insensitive file systems like HFS+ on Mac OS X i.e. macOS and Microsoft Windows filesystems VFAT and FAT32.

There seems to be an issue in regard of handling the current _inc/Plugin.php_ (capital P) file and the previous _inc/plugin.php_ file since the _Release 2018-04-22 "Greebo"_ dokuwiki upgrade.
- Stable 2018-04-22 "Greebo" contains a file _inc/Plugin.php_ (capital P) by 2018-04-22 with 251 Byte.
- Old stable 2017-02-19e "Frusterick Manners" contains a file _inc/plugin.php_ by 2017-08-23 with 10,023 Byte.

with the current stable **deleted.files** list referring the previous _inc/plugin.php_ file to be removed:
```
# This is a list of files that were present in previous DokuWiki releases
# but were removed later. An up to date DokuWiki should not have any of
# the files installed

# removed in 2018-04-22
data/security.png
data/security.xcf
inc/EmailAddressValidator.php
inc/blowfish.php
inc/feedcreator.class.php
inc/lessc.inc.php
inc/plugin.php
lib/images/loading.gif
lib/tpl/dokuwiki/css/_search.css
vendor/easybook/geshi
vendor/phpseclib/phpseclib/composer.lock

# remove in 2017-02-19
```
see (https://raw.githubusercontent.com/splitbrain/dokuwiki/stable/data/deleted.files)

Please have a look at:
1.  the discussion in the [DokuWiki Forum 60569: "After upgrade to Greebo, failed to open stream"](https://forum.dokuwiki.org/post/60569)
2. the issue no. 5 [Possible issue with removal of "inc/Plugin.php" since "Greebo"](https://github.com/Taggic/removeold/issues/5) of the  _removeold_ plugin
3. [Case sensitivity](https://en.wikipedia.org/wiki/Case_sensitivity)

Presumably a name change like _inc/Plugin**s**.php_ (adding an "s") to the current _inc/Plugin.php_ file would be sufficient. This name change and the hopefully minor changes in the calling files could be included in a future _Release 2018-04-22**a** "Greebo"_ dokuwiki upgrade.

**Comment by micgro42** [comment]
Shouldn't this problem be effectively resolved with the updated upgrade plugin in https://github.com/splitbrain/dokuwiki-plugin-upgrade/commit/7b7da9aa0b4871c2238e7ba2388f2f63c61c5f17 ?

**Comment by TP75** [comment]
IMHO there should be a more broader view including easy updates by copying tar-balls.

Please have a look at both [dokuwiki.org: Upgrading DokuWiki](https://www.dokuwiki.org/install:upgrade) and [dokuwiki.org: Removing Unused Files](https://www.dokuwiki.org/install:unused_files).

This reads:
> DokuWiki is very simple to upgrade. Its file layout is organized in a way that avoids overwriting any user content, which means an upgrade generally is just unpacking a new version on top of an old one. You should first try the simple instructions given here, if you encounter problems or the instructions are not detailed enough for you, read the details in the section below. Mac users should use Terminal; copying files via Finder will delete all data. 
> ...
> * Step 5: Delete all unused files

and:
> You can also see the current list at https://github.com/splitbrain/dokuwiki/raw/stable/data/deleted.files
>
> Using RemoveOld Plugin
> To remove old files from DokuWiki's Admin menu you can also use the RemoveOld plugin.

From my long years experience with DokuWiki on different platforms both the **simple upgrade** and the quite reliable **remove old files** procedure are main features for the usability and resilience of a DokuWiki installation. I hope you could agree.

**Comment by splitbrain** [comment]
Yeah, I agree that the current situation is a bit messed up. The problem is, that case insensitive file systems will not change the casing when an existing file is overwritten.

This means when a `plugin.php` exists, extracting a `Plugin.php` from an archive will unpack the file as `plugin.php`. Then the user checks the list of deleted files, sees that `plugin.php` ought to be deleted and does that.

The user does everything right, they delete a file that ought not to exist. Only that this was actually a file that should have a different name altogether. This is a shitty situation.

What we could do is remove `plugin.php` from `deleted.files`. But then users on case sensitive file systems end up with `Plugin.php` and `plugin.php`. It's not the end of the world, but it's kinda messy, too. However it will not crash the user's system.

We probably can not avoid having case renames in the future, because we're in the process of adopting PSR-2, which says classes have to begin with an upper case letter and the file name needs to match the class name.

So I guess what I'm saying is we should probably remove `plugin.php` from the list of deleted files and do not add future case renames as deletions. This leaves dangling files on case sensitive systems but it's probably the better overall solution.

**Comment by mforkel** [comment]
If the upgrade is done by copying files from a tarball generated with the option _Download changed files only_ on [Download DokuWiki!](https://download.dokuwiki.org/), then may be that tarball could include a list of files to be deleted that is relative to the last stable version ('data/deleted.files' includes all files removed since 2006-03-09). The files on that relative list could then be deleted before copying in the new files. That would allow for case changes in file names on case-insensitive file systems and would also allow to reuse file names that were removed in earlier versions.

Also, scripts used to delete the listed files should be case aware. At least the Python script on [Removing Unused Files](https://www.dokuwiki.org/install:unused_files) currently is not. Here is a version that is case sensitive (and will also delete directories included in the list):

```
import os
import shutil

def exists_casesensitive(path):
    if not os.path.exists(path):
        return False
    directory, filename = os.path.split(path)
    return filename in os.listdir(directory)

with open("./data/deleted.files") as file:
    for line in file:
        line = line.strip()
        if line and not line.startswith('#'):
            path = line.rstrip(os.linesep)
            if exists_casesensitive(path):
                if os.path.isdir(path):
                    shutil.rmtree(path)
                    print('Directory removed =>  ' + path)
                else:
                    os.remove(path)
                    print('File removed =>  ' + path)
            else:
                #print('File not found => ' + path)
                pass
```

**Comment by splitbrain** [comment]
Thanks for the python script. Can you update the wiki please?

Generally deleting the files first, then updating will work. (Regardless of choosing updated files only or if the list contains all the deleted file ever or just the ones of the most recent change).
